src/hotspot/share/gc/g1/g1ConcurrentRefineThread.cpp

```diff
@@ -151,3 +105,164 @@ void G1ConcurrentRefineThread::run_service() {
+}
+
+// The primary refinement thread is notified when buffers/cards are added to
+// the dirty card queue.  That can happen in fairly arbitrary contexts.
+// This means there may be arbitrary other locks held when notifying.  We
+// also don't want to have to take a lock on the fairly common notification
+// path, as contention for that lock can significantly impact performance.
+//
+// We use a semaphore to implement waiting and unblocking, to avoid
+// lock rank checking issues.  (We could alternatively use an
+// arbitrarily low ranked mutex.)  The atomic variable _threshold is
+// used to decide when to signal the semaphore.  When its value is
+// SIZE_MAX then the thread is running.  Otherwise, the thread should
+// be requested to run when notified that the number of cards has
+// exceeded the threshold value.
+
+bool G1PrimaryConcurrentRefineThread::wait_for_completed_buffers() {
+  assert(this == Thread::current(), "precondition");
+  _notifier.wait();
+  assert(Atomic::load(&_threshold) == SIZE_MAX || should_terminate(), "incorrect state");
+  return !should_terminate();
+}
+
+bool G1PrimaryConcurrentRefineThread::maybe_deactivate() {
+  assert(this == Thread::current(), "precondition");
+  assert(Atomic::load(&_threshold) == SIZE_MAX, "incorrect state");
+  Atomic::store(&_threshold, cr()->primary_activation_threshold());
+  // Always deactivate when no refinement work found.  New refinement
+  // work may have arrived after we tried, but checking for that would
+  // still be racy.  Instead, the next time additional work is made
+  // available we'll get reactivated.
+  return true;
+}
+
+void G1PrimaryConcurrentRefineThread::activate() {
+  assert(this != Thread::current(), "precondition");
+  // The thread is running when notifications are disabled, so shouldn't
+  // signal is this case.  But there's a race between stop requests and
+  // maybe_deactivate, so also signal if stop requested.
+  size_t threshold = Atomic::load(&_threshold);
+  if (((threshold != SIZE_MAX) &&
+       (threshold == Atomic::cmpxchg(&_threshold, threshold, SIZE_MAX))) ||
+      should_terminate()) {
+    _notifier.signal();
+  }
+}
+
+void G1PrimaryConcurrentRefineThread::notify(size_t num_cards) {
+  // Only activate if the number of pending cards exceeds the activation
+  // threshold.  Notification is disabled when the thread is running, by
+  // setting _threshold to SIZE_MAX.  A relaxed load is sufficient; we don't
+  // need to be precise about this.
+  if (num_cards > Atomic::load(&_threshold)) {
+    // Discard notifications occurring during a safepoint.  A GC safepoint
+    // may dirty some cards (such as during reference processing), possibly
+    // leading to notification.  End-of-GC update_notify_threshold activates
+    // the primary thread if needed.  Non-GC safepoints are expected to
+    // rarely (if ever) dirty cards, so defer activation to a post-safepoint
+    // notification.
+    if (SafepointSynchronize::is_at_safepoint()) {
```

Comment:
@walulyai noticed that this test was backward.  Difference in testing is often surprisingly subtle, I suspect because there's enough other wonkiness in the refinement thread control to often mask the effect.  This change as a whole is intended as a step toward fixing that.  I'm sure glad Ivan spotted this now though.

Reply:
I re-ran tests and ran some additional with this mistake fixed.  No problems encountered.  I also did some better performance testing.  specjbb2015 with options that will trigger the need for more concurrent refinement was unchanged between baseline and with the above fixed.  Having the test backward produced a 6-7% regression.  Thanks again Ivan.